Source/Urho3D/Network/HttpRequest.cpp

```diff
@@ -41,41 +41,140 @@ namespace Urho3D
+    {
+        LogFetch("HTTP OnFetchSucceeded", fetch);
+
+        HttpRequest* request = static_cast<HttpRequest*>(fetch->userData);
+        MutexLock lock(request->mutex_);
+        request->state_ = HTTP_CLOSED;
+        request->readBuffer_.Resize(fetch->numBytes - 1);
+        request->readBuffer_.SetData(fetch->data, fetch->numBytes - 1);
+        request->requestHandle_ = nullptr;
+
+        emscripten_fetch_close(fetch);
+    };
+
+    attr.onerror = [](emscripten_fetch_t* fetch)
+    {
+        LogFetch("HTTP OnFetchError", fetch);
+
+        HttpRequest* request = static_cast<HttpRequest*>(fetch->userData);
+        MutexLock lock(request->mutex_);
+        request->state_ = HTTP_ERROR;
+        request->error_ = fetch->statusText;
+        request->requestHandle_ = nullptr;
+
+        emscripten_fetch_close(fetch);
+    };
+
+    attr.onprogress = [](emscripten_fetch_t * fetch)
+    {
+        LogFetch("HTTP OnFetchProgress", fetch);
+    };
+
+    attr.onreadystatechange = [](emscripten_fetch_t * fetch)
+    {
+        LogFetch("HTTP OnFetchReadyStateChange", fetch);
+    };
+
+    ea::vector<const char*> requestHeaders;
+    ea::vector<ea::string> requestHeadersStr;
+    if (!headers_.empty())
+    {
+        for (const auto& header : headers_)
         {
-            HttpRequest* request = static_cast<HttpRequest*>(arg);
-            MutexLock lock(request->mutex_);
-            request->state_ = HTTP_ERROR;
-            request->error_ = msg;
-            request->requestHandle_ = 0;
-        }, [](unsigned handle, void* arg, int loaded, int total)        // onprogress
+            const auto pair = header.trimmed().split(':');
+
+            const auto expectedSize = 2;
+            if (pair.size() != expectedSize)
+            {
+                URHO3D_LOGWARNING("HTTP ignoring header '{}' with unexpected format, expected format 'key: value'", header);
+                continue;
+            }
+
+            const auto keyIndex = 0;
+            const auto& key = pair[keyIndex].trimmed();
+            if (key.empty())
+            {
+                URHO3D_LOGWARNING("HTTP ignoring header '{}' with empty key", header);
+                continue;
+            }
+
+            const auto valueIndex = 1;
+            const auto& value = pair[valueIndex].trimmed();
+            if (value.empty())
+            {
+                URHO3D_LOGWARNING("HTTP ignoring header '{}' with empty value", header);
+                continue;
+            }
+
+            requestHeadersStr.push_back(key);
+            requestHeadersStr.push_back(value);
+        }
+
+        if (!requestHeadersStr.empty())
         {
-            //HttpRequest* request = static_cast<HttpRequest*>(arg);
-        });
+            for (const auto& rh : requestHeadersStr)
+            {
+                requestHeaders.push_back(rh.c_str());
+            }
+            requestHeaders.push_back(nullptr);
+            attr.requestHeaders = requestHeaders.data();
```

Comment:
I am not sure it is safe to pass the pointer to _local_ variable here. I think that fetch can and should usually be async. It can be fixed by either making the fetch sync (which probably sucks), or by using members.